### PR TITLE
fix for glpi10 with change in updateActor in postUpdateItem the plugi…

### PR DESCRIPTION
**Pull request by ales-infotel**
…n remove a group, then Add the new group from Category then Remove group and add the first group
fix the issue #119

### CODE REVIEW

File: inc/ticket.class.php [hunk 1]
```diff
@@ -323,7 +323,7 @@ static function addHistoryOnAddGroup(CommonDBTM $item) {
       return $item;
    }
 
-   static function processAfterAddGroup(CommonDBTM $item) {
+   static function processAfterAddGroup(CommonDBTM &$item) {
```
Comment (cedric-anne) [inline inc/ticket.class.php:326]: In PHP, objects are always passed by reference.
```suggestion
   static function processAfterAddGroup(CommonDBTM $item) {
```

**Review by cedric-anne — COMMENTED** [review]
I do not really know how this plugin should work, so I am not sure I can do a functionnal review of such changes. I would prefer to have @stonebuzz validation before doing a technical review.

**Comment by CorentinS6** [comment]
ping @stonebuzz

**Review by stonebuzz — CHANGES_REQUESTED** [review]
All changes can be simplified as follows (IMHO and and after having done several tests)

```patch
diff --git a/inc/ticket.class.php b/inc/ticket.class.php
index 793be33..8c77ade 100644
--- a/inc/ticket.class.php
+++ b/inc/ticket.class.php
@@ -642,9 +642,18 @@ class PluginEscaladeTicket {
          ];
          $group_found = $group_ticket->find($group_condition);
          if (empty($group_found)) {
-
             //add group to ticket
             $group_ticket->add($group_condition);
+            //remove old group if needed
+            if ($_SESSION['plugins']['escalade']['config']['remove_group']) {
+               if(isset($item->input['_groups_id_assign'])) {
+                  foreach ($item->input['_groups_id_assign'] as $idActor => $actor) {
+                     unset($item->input['_groups_id_assign'][$idActor]);
+                     unset($item->input['_groups_id_assign_notif']['use_notification'][$idActor]);
+                     unset($item->input['_groups_id_assign_notif']['alternative_email'][$idActor]);
+                  }
+               }
+            }
          }
       }
 
@@ -662,9 +671,18 @@ class PluginEscaladeTicket {
          ];
          $user_found = $ticket_user->find($user_condition);
          if (empty($user_found)) {
-
             //add user to ticket
             $ticket_user->add($user_condition);
+            //remove old tech if needed
+            if($_SESSION['plugins']['escalade']['config']['remove_tech']) {
+               if(isset($item->input['_users_id_assign'])) {
+                  foreach ($item->input['_users_id_assign'] as $idActor => $actor) {
+                     unset($item->input['_users_id_assign'][$idActor]);
+                     unset($item->input['_users_id_assign_notif']['use_notification'][$idActor]);
+                     unset($item->input['_users_id_assign_notif']['alternative_email'][$idActor]);
+                  }
+                }
+            }
          }
       }
    }

```

I tried to understand the current behavior.

After adding the category group, the plugin process reaches the step of deleting the old group.

until then "normal"

Here is the full stack trace of this step

```shell
#0 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/escalade/inc/ticket.class.php(335): PluginEscaladeTicket::removeAssignGroups()
#1 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/escalade/hook.php(430): PluginEscaladeTicket::processAfterAddGroup()
#2 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/Plugin.php(1617): plugin_escalade_item_add_group_ticket()
#3 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/CommonDBTM.php(1364): Plugin::doHook()
#4 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/escalade/inc/ticket.class.php(658): CommonDBTM->add()
#5 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/escalade/inc/ticket.class.php(83): PluginEscaladeTicket::qualification()
#6 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/escalade/hook.php(367): PluginEscaladeTicket::item_update()
#7 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/Plugin.php(1617): plugin_escalade_item_update()
#8 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/CommonDBTM.php(1678): Plugin::doHook()
#9 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/front/ticket.form.php(84): CommonDBTM->update()
```

After deleting the old group, the ticket is updated (at least the modification date)  which trigger ```CommonITILObject->updateActors()```

Here is the full stack trace of this second step

```shell
#0 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/escalade/inc/ticket.class.php(335): PluginEscaladeTicket::removeAssignGroups()
#1 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/plugins/escalade/hook.php(430): PluginEscaladeTicket::processAfterAddGroup()
#2 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/Plugin.php(1617): plugin_escalade_item_add_group_ticket()
#3 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/CommonDBTM.php(1364): Plugin::doHook()
#4 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/CommonITILObject.php(8309): CommonDBTM->add()
#5 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/CommonITILObject.php(1910): CommonITILObject->updateActors()
#6 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/Ticket.php(1625): CommonITILObject->post_updateItem()
#7 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/src/CommonDBTM.php(1719): Ticket->post_updateItem()
#8 /home/stonebuzz/TECLIB/DEV/GLPI/10.0-bugfixes/front/ticket.form.php(84): CommonDBTM->update()
```

```CommonITILObject->updateActors()``` seems to remember the old group, but can't find it since the DB, so it adds it.

Functionally it is OK for me, the behavior is restored.

Technically I don't understand why ```unset``` works, because ```$item``` is not passed by reference (but it seems to be the case).

**Comment by stonebuzz** [comment]
The @cedric-anne  review remains valid

**Comment by CorentinS6** [comment]
ping @ales-infotel

File: inc/ticket.class.php [hunk 2]
```diff
@@ -643,9 +643,20 @@ static function qualification(CommonDBTM $item) {
          $group_found = $group_ticket->find($group_condition);
          if (empty($group_found)) {
 
-            //add group to ticket
-            $group_ticket->add($group_condition);
+             //add group to ticket
+             $group_ticket->add($group_condition);
+             +            //remove old group if needed
```
Comment (cedric-anne) [inline inc/ticket.class.php:648]:
```suggestion
             //remove old group if needed
```

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**